### PR TITLE
Lab4: Implement numeronym solution and tests

### LINKED ISSUE #523: Lab 4 - Numeronym (alanhillgemann)

### PR DESCRIPTION
**Pull request by alanhillgemann**
Fixes #523 

Review of colleague's PR #tba

### CODE REVIEW

File: 04_numeronym/alanhillgemann/main.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+)
+
+var out io.Writer = os.Stdout
+
+func main() {
+	fmt.Fprint(out, numeronyms("accessibility", "Kubernetes", "abc"))
+}
+
+func numeronyms(vals ...string) []string {
+
```
Comment (anzdaddy) [inline 04_numeronym/alanhillgemann/main.go]: Unnecessary blank line

File: 04_numeronym/alanhillgemann/main.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	numeronyms := make([]string, 0)
+	for _, val := range vals {
+		if val == "" {
+			numeronyms = append(numeronyms, "")
+		} else {
+			strArr := []rune(val)
```
Comment (anzdaddy) [inline 04_numeronym/alanhillgemann/main.go]: Name could be more descriptive. Maybe `runes`?

File: 04_numeronym/alanhillgemann/main.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+			staVal := string(strArr[0])
```
Comment (anzdaddy) [inline 04_numeronym/alanhillgemann/main.go]: Not obvious what `staVal` means. Perhaps `firstVal` is better.

Comment (anzdaddy) [inline 04_numeronym/alanhillgemann/main.go]: Might `val` be empty?

File: 04_numeronym/alanhillgemann/main.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+			midVal := ""
+			endVal := ""
+			switch {
+			case len(strArr) == 2:
+				endVal = string(strArr[len(strArr)-1])
```
Comment (anzdaddy) [inline 04_numeronym/alanhillgemann/main.go]: You'll probably find the code works out simpler if you compute `endVal` outside the `switch` under its own `if`.

File: 04_numeronym/alanhillgemann/main.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+	numeronyms := make([]string, 0)
+	for _, val := range vals {
+		if val == "" {
+			numeronyms = append(numeronyms, "")
+		} else {
+			strArr := []rune(val)
+			staVal := string(strArr[0])
+			midVal := ""
+			endVal := ""
+			switch {
```
Comment (anzdaddy) [inline 04_numeronym/alanhillgemann/main.go]:
```go
switch len(strArr) {
case 2:
```

**Review by anzdaddy — CHANGES_REQUESTED** [review]

File: 04_numeronym/alanhillgemann/main.go (hunk 1 above)
Comment (alanhillgemann) [reply 04_numeronym/alanhillgemann/main.go]: Fixed

File: 04_numeronym/alanhillgemann/main.go (hunk 2 above)
Comment (alanhillgemann) [reply 04_numeronym/alanhillgemann/main.go]: Fixed. All variables renamed to be more descriptive

File: 04_numeronym/alanhillgemann/main.go (hunk 3 above)
Comment (alanhillgemann) [reply 04_numeronym/alanhillgemann/main.go]: Fixed. All variables renamed to be more descriptive

Comment (alanhillgemann) [reply 04_numeronym/alanhillgemann/main.go]: Yes, handled on row 18

File: 04_numeronym/alanhillgemann/main.go (hunk 4 above)
Comment (alanhillgemann) [reply 04_numeronym/alanhillgemann/main.go]: Fixed, switch replaced with if statements

File: 04_numeronym/alanhillgemann/main.go (hunk 5 above)
Comment (alanhillgemann) [reply 04_numeronym/alanhillgemann/main.go]: Fixed, switch replaced with if statements

**Comment by alanhillgemann** [comment]
@anzdaddy thank you for reviewing my PR. I have made all changes requested.

File: 04_numeronym/alanhillgemann/main.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+			firstChar := string(runes[0])
+			middleChar := ""
+			lastChar := ""
+			if len(runes) > 1 {
+				lastChar = string(runes[len(runes)-1])
+			}
+			if len(runes) == 3 {
+				middleChar = string(runes[1])
+			}
+			if len(runes) > 3 {
```
Comment (anzdaddy) [inline 04_numeronym/alanhillgemann/main.go]: If you change this to `>= 3` or `> 2`, it covers the second case (`== 3`).

**Review by anzdaddy — CHANGES_REQUESTED** [review]

File: 04_numeronym/alanhillgemann/main.go (hunk 6 above)
Comment (alanhillgemann) [reply 04_numeronym/alanhillgemann/main.go]: Improved by nesting if statements

**Comment by alanhillgemann** [comment]
@anzdaddy thank you for reviewing my PR again. I have made all changes requested.

**Review by anzdaddy — APPROVED** [review]

**Pull request merged**